src/components/route-book/index.tsx

```diff
@@ -0,0 +1,154 @@
+import { useBook } from '@/fetchers/book';
+import { fromBaseUnit } from '@/old/utils/math/hiLo';
+import { Metadata } from '@penumbra-zone/protobuf/penumbra/core/asset/v1/asset_pb';
+import {
+  Position,
+  PositionId,
+} from '@penumbra-zone/protobuf/penumbra/core/component/dex/v1/dex_pb';
+import { Amount } from '@penumbra-zone/protobuf/penumbra/core/num/v1/num_pb';
+import { getDisplayDenomExponent } from '@penumbra-zone/getters/metadata';
+import { innerToBech32Address } from '@/old/utils/math/bech32';
+import { uint8ArrayToBase64 } from '@/old/utils/math/base64';
+import { JsonValue } from '@bufbuild/protobuf';
+import { useAssets } from '@/shared/state/assets';
+import { round } from '@/shared/round';
+import { useComputePositionId } from '@/shared/useComputePositionId';
+
+interface Route {
+  lpId: string;
+  position: Position;
+  price: number;
+  amount: number;
+}
+
+interface RouteWithTotal extends Route {
+  total: number;
+}
+
+function getTotals(data: Route[], isBuySide: boolean, limit: number): RouteWithTotal[] {
+  const totals: RouteWithTotal[] = Object.values(
+    data.reduce((displayData: Record<string, RouteWithTotal>, route: Route) => {
+      const key = route.price;
+      return {
+        ...displayData,
+        [key]: displayData[key]
+          ? {
+              ...route,
+              total: displayData[key].total + route.amount,
+              amount: Math.min(displayData[key].amount, route.amount),
+            }
+          : {
+              ...route,
+              total: route.amount,
+            },
+      };
+    }, {}),
+  );
+
+  return isBuySide
+    ? totals.slice(0, limit).sort((a, b) => b.price - a.price)
+    : totals.slice(-limit);
+}
+
+function getDisplayData(
+  data: Position[],
+  computePositionId: ((position: Position) => PositionId) | undefined,
+  asset1: Metadata | undefined,
+  asset2: Metadata | undefined,
+  isBuySide: boolean,
+  limit: number,
+): RouteWithTotal[] {
+  if (!computePositionId || !asset1 || !asset2) {
+    return [];
+  }
+
+  const routes = data
```

Comment:
question: where did all the domain knowledge for the routes come from? Did it come from the core repo? Can we add documentation on these steps or code comments to help the next person?